### PR TITLE
Add CHANGELOG.md for ofrak, ofrak-io, ofrak-type, ofrak-patch-maker

### PR DESCRIPTION
**Pull request by whyitfor**
**Please describe the changes in your request.**
Add changelogs for ofrak, ofrak-io, ofrak-type, ofrak-patch-maker.

As the changelogs indicate, they are bassed off of [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).

Specifically, there are 5 types of changes that are logged:

- `Added` for new features.
- `Changed` for changes in existing functionality.
- `Deprecated` for soon-to-be removed features.
- `Removed` for now removed features.
- `Fixed` for any bug fixes.
- `Security` in case of vulnerabilities.

This PR also updates `CONTRIBUTING.md` to mention changelogs and test coverage as part of Pull Request process.

This may not be perfect, but seems like a reasonable starting point.

**Anyone you think should look at this, specifically?**
@andresito00, @EdwardLarson, @rbs-jacob.

### CODE REVIEW

File: ofrak_core/CHANGELOG.md [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+### Fixed
+- `Resource.create_child` takes either `data` or `data_range` arguments, but not both.
+- Remove duplicated work from in resource service serialization, making it faster.
+- Miscellaneous bugs in documentation.
+- Added `beartype` as requirement for `ofrak`.
+- Bump to `fun-coverage==0.2.0` for more accurate test coverage reporting.
+
+### Security
+in case of vulnerabilities.
```
Comment (rbs-jacob) [inline ofrak_core/CHANGELOG.md:49]: Is this line supposed to be here?

```suggestion
```

File: ofrak_core/CHANGELOG.md [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+- Free space components decoupled from ISA
+- `Filesystem.initialize_from_disk` raises `FileNotFoundError` when path does not exist or is not a file ([@galbwe](https://github.com/galbwe))
+- Rename `RawExtendModifier` to `BinaryExtendModifier`
+- `ElfProgramHeader` simplified to have only one method, `get_memory_permissions`.
+- Switch to using PyPI versions of keystone & capstone.
+- `PeOptionalHeader` divided into 32 and 64 bit.
+
+### Removed
+- Removed unused, untested APIs:
+  - Removed `Resource.{get_data_index_within_parent, get_offset_within_root, get_data_unmapped_range, set_data_alignment, fetch, get_related_tags, get_all_attributes, move, get_siblings_as_view, get_siblings`.
```
Comment (rbs-jacob) [inline ofrak_core/CHANGELOG.md:34]: Missing trailing `}`.

```suggestion
  - Removed `Resource.{get_data_index_within_parent, get_offset_within_root, get_data_unmapped_range, set_data_alignment, fetch, get_related_tags, get_all_attributes, move, get_siblings_as_view, get_siblings}`.
```

File: ofrak_core/CHANGELOG.md [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+- `ofrak` now contains all the components that were in the (now sunset) `ofrak-components` package.
+- Add components for analyzing, unpacking, and repacking raw flash dumps.
+- UF2 unpacker and packer added.
+- Several new features to make tracking and installing external dependencies easier (for non-Docker users):
+  - Components can now mark their external tool dependencies using the `external_dependencies` attribute.
+  - CLI tool added to track and install dependencies & components:
+    - `python -m ofrak list` lists installed OFRAK modules and/or components.
+    - `python -m ofrak deps` shows and checks external (non-Python) dependencies of OFRAK components, and shows hints on how they might be installed.
+  - `OFRAKContext` logs if a discovered component is missing an external dependency; these components can be excluded using the `exclude_components_missing_depenencies` parameter in `OFRAK`.
+- Speed improvements to `resource.unpack_recursively` method.
```
Comment (rbs-jacob) [inline ofrak_core/CHANGELOG.md:19]:
```suggestion
- Speed improvements to `Resource.unpack_recursively` method.
```

File: ofrak_core/CHANGELOG.md [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+- `DataService` was refactored for speed and clarity of implementation.
+
+### Changed
+- `Resource.{write_to, flush_to_disk}` now have optional `pack` parameter to pack before writing
+- Resource views are now updated when `Resource.save` is run
+- Free space components decoupled from ISA
+- `Filesystem.initialize_from_disk` raises `FileNotFoundError` when path does not exist or is not a file ([@galbwe](https://github.com/galbwe))
+- Rename `RawExtendModifier` to `BinaryExtendModifier`
```
Comment (rbs-jacob) [inline ofrak_core/CHANGELOG.md:27]: These should probably end with periods for consistency with the rest of this document.

Comment (rbs-jacob) [reply ofrak_core/CHANGELOG.md:27]: Worth doing a pass over all the bullets to make sure they're consistent – there are a few other places lower in the document that are also missing consistent trailing punctuation.

File: ofrak_patch_maker/CHANGELOG.md [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/) and adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
+
+## [Unreleased](https://github.com/redballoonsecurity/ofrak/tree/master)
+
+## [2.0.0](https://github.com/redballoonsecurity/ofrak/releases/tag/ofrak-patch-maker-v.2.0.0) - 2023-01-03
+### Changed
+- `Toolchain` interface uses `ArchInfo` instead of `ProgramAttributes` to remove dependency on `ofrak`.
+- `bss_size_required` parameter added to `AssembledObject`.
+- `segment_alignment` parameter added to `BOM`.
+- Removed reference to obsolete `ofrak-componets` from README.md.
```
Comment (rbs-jacob) [inline ofrak_patch_maker/CHANGELOG.md:13]:
```suggestion
- Removed reference to obsolete `ofrak_components` from README.md.
```

File: ofrak_type/CHANGELOG.md [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+
+## [Unreleased](https://github.com/redballoonsecurity/ofrak/tree/master)
+
+## [2.0.0](https://github.com/redballoonsecurity/ofrak/releases/tag/ofrak-type-v2.0.0) - 2023-01-03
+### Added
+- `ArchInfo` to describe an architecture
+- Classes in package all discoverable at `ofrak_type`, e.g., `from ofrak_type import Range`.
+
+### Changed
+- Removed reference to obsolete `ofrak-components` from README.md.
```
Comment (rbs-jacob) [inline ofrak_type/CHANGELOG.md:14]:
```suggestion
- Removed reference to obsolete `ofrak_components` from README.md.
```

File: ofrak_core/CHANGELOG.md [hunk 7]
```diff
@@ -0,0 +1,53 @@
+# Changelog
+All notable changes to `ofrak` will be documented in this file.
+
+The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/) and adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
+
+## [Unreleased](https://github.com/redballoonsecurity/ofrak/tree/master)
+
+## [2.0.0](https://github.com/redballoonsecurity/ofrak/releases/tag/ofrak-v2.0.0) - 2023-01-03
+### Added
+- `ofrak` now contains all the components that were in the (now sunset) `ofrak-components` package.
```
Comment (rbs-jacob) [inline ofrak_core/CHANGELOG.md:10]:
```suggestion
- `ofrak` now contains all the components that were in the (now sunset) `ofrak_components` package.
```

File: ofrak_io/CHANGELOG.md [hunk 8]
```diff
@@ -0,0 +1,15 @@
+# Changelog
+All notable changes to `ofrak-io` will be documented in this file.
+
+The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/) and adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
+
+## [Unreleased](https://github.com/redballoonsecurity/ofrak/tree/master)
+
+## [1.1.0](https://github.com/redballoonsecurity/ofrak/releases/tag/ofrak-io-v1.1.0) - 2023-01-03
+### Changed
+- Removed reference to obsolete `ofrak-components` from README.md.
```
Comment (rbs-jacob) [inline ofrak_io/CHANGELOG.md:10]:
```suggestion
- Removed reference to obsolete `ofrak_components` from README.md.
```

**Review by rbs-jacob — CHANGES_REQUESTED** [review]

File: ofrak_core/CHANGELOG.md (hunk 1 above)
Comment (whyitfor) [reply ofrak_core/CHANGELOG.md:49]: Good catch!

File: ofrak_core/CHANGELOG.md (hunk 4 above)
Comment (whyitfor) [reply ofrak_core/CHANGELOG.md]: Good catch, I went through all the docs again as well.

**Review by rbs-jacob — APPROVED** [review]

File: ofrak_type/CHANGELOG.md [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+## [2.0.0](https://github.com/redballoonsecurity/ofrak/releases/tag/ofrak-type-v2.0.0) - 2023-01-03
+### Added
+- Add `ArchInfo` to describe an architecture.
+- Classes in package all discoverable at `ofrak_type`, e.g., `from ofrak_type import Range`.
+
+### Changed
+- Removed reference to obsolete `ofrak_components` from README.md.
+
+### Fixed
+- Fixed bug that flipped integer values of  `MemoryPermissions.R` and `MemoryPermission.X`.
```
Comment (EdwardLarson) [inline ofrak_type/CHANGELOG.md:17]: I consider this a "Change" - it's the reason why the major version changed.

Comment (whyitfor) [reply ofrak_type/CHANGELOG.md:18]: Ok -- I added it in `Change` as well.

File: ofrak_type/CHANGELOG.md [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+### Added
+- Add `ArchInfo` to describe an architecture.
+- Classes in package all discoverable at `ofrak_type`, e.g., `from ofrak_type import Range`.
+
+### Changed
+- Removed reference to obsolete `ofrak_components` from README.md.
+- Integer values for `MemoryPermission.R` and `MemoryPermission.X`now correctly match bitfield values.
+
+### Fixed
+- Fixed bug that flipped integer values of  `MemoryPermissions.R` and `MemoryPermission.X`.
```
Comment (EdwardLarson) [inline ofrak_type/CHANGELOG.md:18]:
```suggestion
- Integer values for `MemoryPermission.R` and `MemoryPermission.X` now correctly match bitfield values.
```

**Pull request merged**